### ISSUE TITLE
`children` prop should be callable to add attributes

### ISSUE THREAD

**Issue by Blazzike** [solidjs/solid#1605]
As far as I can tell, there is no easy way of adding attributes to children passed into the parent.

Adding them to a client side render is easy enough, but for SSR beyond parsing the output HTML, I'm not sure how you'd do it.

Currently accessing the children prop results in:
```js
get children() {
  return _$ssrElement("tag", {
    /* attributes */
  }, undefined, true);
}
```

Instead or rather additionally as not to break compatibility the implementation should look something like the following:
```js
children(...additionalAttributes) {
  return _$ssrElement("tag", {
    /* attributes */
    ...additionalAttributes,
  }, undefined, true);
}
```

So it can be used like so:
```jsx
<my-amazing-wrapper>
  {children({
    'data-status': 'OK',
  })}

  {/* or */}
  {children()}

  {/* or */}
  {children}
</my-amazing-wrapper>
```

Cheers!

**Comment by thetarnav** [comment]
I would be great to have some official primitive in solid/dom-expressions for that use.

Kobalte-ui is solving that through a polymorphic `<As component="button" {...props}>` component and a [`combineProps`](https://github.com/solidjs-community/solid-primitives/tree/main/packages/props#combineprops) primitive. But it requires the component to be passed to `As` as a factory function, rather than a resolved dom node: https://github.com/kobaltedev/kobalte/blob/main/packages/core/src/polymorphic/polymorphic.tsx

There is also an interesting experiment in solid-radix for applying attributes to already resolved elements. (like `spread` that combines attributes instead of overriding them) https://github.com/TrentsPC/solid-radix/blob/main/src/solid/Slot.tsx
But it doesn't work in SSR

**Comment by ryansolid** [comment]
Yeah it's is tricky because at the time you get them they've already been created. On the client yeah you can modify the DOM. On the server that would basically be string manipulation. Truthfully child introspection is a really awkward pattern for declarative libraries and gets in the way of optimization. Even the React Core team (in React this pattern is more common) considers this a bad pattern and effectively deprecated in their minds. I get the scenarios and the APIs it opens up but I think it requires careful consideration.

Stuff like passing in the factory function is a better pattern in general. I think I'm going to move this into discussions so we can collect ideas here.
